drivers/wifi/simplelink.c

```diff
@@ -0,0 +1,183 @@
+/**
+ * Copyright (c) 2018 Texas Instruments, Incorporated
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_WIFI_LEVEL
+#define SYS_LOG_DOMAIN "dev/simplelink"
+#include <logging/sys_log.h>
+
+#include <zephyr.h>
+#include <kernel.h>
+#include <device.h>
+#include <net/net_if.h>
+#include <net/wifi_mgmt.h>
+#include <net/net_offload.h>
+
+#include "simplelink_support.h"
+
+struct simplelink_data {
+	struct net_if *iface;
+	unsigned char mac[6];
+};
+
+static struct simplelink_data simplelink_data;
+
+/* Handle connection events from the SimpleLink Event Handlers: */
+static void simplelink_wifi_cb(u32_t mgmt_event, struct sl_connect_state *conn)
+{
+	struct in_addr addr;
+	struct in_addr gwaddr;
+	int status;
+
+	/*
+	 * Once Zephyr wifi_mgmt wifi_status codes are defined, will need
+	 * to map from SimpleLink error codes.  For now, just return -EIO.
+	 */
+	status = (conn->error ? -EIO : 0);
+
+	switch (mgmt_event) {
+	case NET_EVENT_WIFI_CMD_CONNECT_RESULT:
+		/* Only get this event if connect succeeds: */
+		wifi_mgmt_raise_connect_result_event(simplelink_data.iface,
+						     status);
+		k_yield(); /* Needed to fire net_event notifications: */
```

Comment:
That's actually not true.

The event will be fired, with or without the k_yield(). You are confusing with the scan result event case where using k_yield() helps to avoid loosing events if the system is over-loaded, knowing that there might be numerous scan result events, that's all. Here it's a lone event, unlikely to be lost.

Remove this

Reply:
> Here it's a lone event, unlikely to be lost.

Actually, during testing, when providing an SSID with an invalid security key, I found that the offload chip continues to retry the connection.   It has an "auto retry" feature.   So, in this case, disconnect events are generated in a higher priority thread than the net mgmt thread, and so the wifi shell never sees any disconnect events.

There may be other similar cases with these offload chips which may not have been originally expected.

But in general, since the net event mgmt system is lossy, there is no guarantee that events will get delivered.  So until net mgmt framework is re-architected to guarantee events are not dropped, I would argue that doing a k_yield() after  each wifi_mgmt_raise_[x]_event() call is mandatory.

So, to be more exact, the comment should say something like "the k_yield() is needed to *ensure* the notification is not dropped."

Reply:
> But in general, since the net event mgmt system is lossy, there is no guarantee that events will get delivered. So until net mgmt framework is re-architected to guarantee events are not dropped, I would argue that doing a k_yield() after each wifi_mgmt_raise_[x]_event() call is mandatory.

No way. net mgmt is meant for user targeted events, so there is just _no point_ in this. This is just a technical non sense.

If you have a solution, please go fix Netlink in Linux as well.

Reply:
Removed.